### PR TITLE
PIM-4037: creation of specs on constraints of the catalog bundle

### PR DESCRIPTION
**Pull request by Yannsimon92**
| Q                    | A
| -------------------- | ---
| Bug fix?             | no
| New feature?         | no
| BC breaks?           | no
| CI currently passes? | yes
| Tests pass?          | yes
| Scenarios pass?      | yes
| Checkstyle issues?*  | no
| PMD issues?**        | no
| Changelog updated?   | no
| Fixed tickets        | PIM-4037
| DB schema updated?   | no
| Migration script?    | no
| Doc PR               | no

### CODE REVIEW

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/ImmutableValidatorSpec.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        $entityManager->getUnitOfWork()->willReturn($unitOfWork);
+        $unitOfWork->getOriginalEntityData($family)->willReturn(['code' => 'MyOriginalCode']);
+
+        $context->addViolationAt(Argument::any(), Argument::any())
+            ->shouldNotBeCalled();
+
+        $constraint->properties = ['code'];
+        $this->validate($family, $constraint);
+    }
+}
```
Comment (willy-ahva) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/ImmutableValidatorSpec.php]: Add a blank line

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeSpec.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        $this->shouldBeAnInstanceOf('Symfony\Component\Validator\Constraint');
+    }
+
+    function it_has_message()
+    {
+        $this->message->shouldBe('An identifier attribute already exists.');
+    }
+}
+
```
Comment (willy-ahva) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeSpec.php]: Remove this one

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
+<?php
+
+namespace spec\Pim\Bundle\CatalogBundle\Validator\Constraints;
+
+use Pim\Bundle\CatalogBundle\Repository\AttributeRepositoryInterface;
+use PhpSpec\ObjectBehavior;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ExecutionContextInterface;
+use Prophecy\Argument;
+
```
Comment (willy-ahva) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php]: Remove this blank line

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+use Pim\Bundle\CatalogBundle\Validator\Constraints\SingleIdentifierAttribute;
+
```
Comment (willy-ahva) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php]: Remove this blank line

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/ImmutableSpec.php [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    function it_is_a_validator_constraint()
+    {
+        $this->shouldBeAnInstanceOf('Symfony\Component\Validator\Constraint');
+    }
+
+    function it_has_properties()
+    {
+        $this->properties->shouldBe(null);
```
Comment (filipsalpe) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/ImmutableSpec.php]: Looks wrong. it_has_no_properties?

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+use Pim\Bundle\CatalogBundle\Model\AttributeInterface;
+
+
```
Comment (filipsalpe) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php]: extra blank line

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        $attribute->getAttributeType()->willReturn('pim_catalog_text');
+
+        $context
+            ->addViolationAt(Argument::cetera())
+            ->shouldNotBeCalled();
+
+        $this->validate($attribute, $constraint);
+    }
+
+     function it_does_nothing_if_identifiers_id_are_the_same(
```
Comment (willy-ahva) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php]: if attribute identifier already exists

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/ImmutableValidatorSpec.php (hunk 1 above)
Comment (Yannsimon92) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/ImmutableValidatorSpec.php]: ok

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php (hunk 3 above)
Comment (Yannsimon92) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php]: ok

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php (hunk 4 above)
Comment (Yannsimon92) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php]: ok

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/ImmutableSpec.php (hunk 5 above)
Comment (Yannsimon92) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/ImmutableSpec.php]: i delete it

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php (hunk 6 above)
Comment (Yannsimon92) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php]: ok

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php (hunk 7 above)
Comment (Yannsimon92) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php]: ok

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/ImmutableValidatorSpec.php [hunk 8]
```diff
@@ -0,0 +1,64 @@
+<?php
+
+namespace spec\Pim\Bundle\CatalogBundle\Validator\Constraints;
+
+use Doctrine\ORM\EntityManager;
+use Doctrine\ORM\UnitOfWork;
+use PhpSpec\ObjectBehavior;
+use Pim\Bundle\CatalogBundle\Entity\Family;
+use Pim\Bundle\CatalogBundle\Model\FamilyInterface;
```
Comment (skeleton) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/ImmutableValidatorSpec.php]: useless

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        $this->shouldHaveType('Pim\Bundle\CatalogBundle\Validator\Constraints\SingleIdentifierAttributeValidator');
+    }
+
+    function it_does_nothing_if_attribute_type_is_not_identifier(
+        $context,
+        AttributeInterface $attribute,
+        Constraint $constraint
+    ) {
+
```
Comment (skeleton) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php]: remove this blank line

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+     function it_does_nothing_if_identifiers_id_are_the_same(
+         $context,
+         $attributeRepository,
+         AttributeInterface $attribute,
+         AttributeInterface $identifier,
+         Constraint $constraint
+     ) {
+
```
Comment (skeleton) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php]: remove this blank line

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    function it_adds_a_violation_if_attribute_identifier_already_exists(
+        $context,
+        $attributeRepository,
+        AttributeInterface $attribute,
+        AttributeInterface $identifier,
+        SingleIdentifierAttribute $constraint
+    ) {
+
```
Comment (skeleton) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php]: remove this blank line

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/ImmutableValidatorSpec.php (hunk 8 above)
Comment (Yannsimon92) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/ImmutableValidatorSpec.php]: ok

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php (hunk 9 above)
Comment (Yannsimon92) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php]: ok

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php (hunk 10 above)
Comment (Yannsimon92) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php]: ok

File: spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php (hunk 11 above)
Comment (Yannsimon92) [inline spec/Pim/Bundle/CatalogBundle/Validator/Constraints/SingleIdentifierAttributeValidatorSpec.php]: ok

File: .idea/php.xml [hunk 12]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="PhpProjectSharedConfiguration" php_language_level="5.5.0" />
```
Comment (nidup) [inline .idea/php.xml]: please don't commit this file

Comment (Yannsimon92) [inline .idea/php.xml]: ok

**Comment by nidup** [comment]
closed in favor of https://github.com/akeneo/pim-community-dev/pull/2612

**Pull request closed without merging**